### PR TITLE
Added Knapsack Problem (recursive Solution)

### PR DESCRIPTION
**Pull request by akashgkrishnan**
Added the recursive implementation of the 01 knapsack problem

# Welcome to Dart community

### **Describe your change:**

* [x] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### **Checklist:**
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Dart/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Dart files are placed inside an existing directory.
* [ ] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
* [ ] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

File: dynamic_programming/01knapsack_recursive.dart [hunk 1]
```diff
@@ -0,0 +1,38 @@
+import 'dart:math';
+import 'package:test/test.dart';
+
+int knapSackProblem(
+    int numberOfItems, int capacity, List<int> values, List<int> weights) {
```
Comment (Parowicz) [inline dynamic_programming/01knapsack_recursive.dart:5]: From what i see you could take numberOfItems from `values.length`
```suggestion
int knapSackProblem(int capacity, List<int> values, List<int> weights,
    [int numberOfItems]) {
  numberOfItems ??= values.length;
```

File: dynamic_programming/01knapsack_recursive.dart [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+  }
+  int currentValue = values[numberOfItems - 1];
+  int currentWeight = weights[numberOfItems - 1];
+
+  if (weights[numberOfItems - 1] <= capacity) {
+    return max(
+        currentValue +
+            knapSackProblem(
+                numberOfItems - 1, capacity - currentWeight, values, weights),
+        knapSackProblem(numberOfItems - 1, capacity, values, weights));
```
Comment (Parowicz) [inline dynamic_programming/01knapsack_recursive.dart:17]:
```suggestion
                capacity - currentWeight, values, weights, numberOfItems - 1),
        knapSackProblem(capacity, values, weights, numberOfItems - 1));
```

File: dynamic_programming/01knapsack_recursive.dart [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+  } else {
+    return knapSackProblem(numberOfItems - 1, capacity, values, weights);
```
Comment (Parowicz) [inline dynamic_programming/01knapsack_recursive.dart:19]:
```suggestion
    return knapSackProblem(capacity, values, weights, numberOfItems - 1);
```

File: dynamic_programming/01knapsack_recursive.dart [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+  }
+}
+
+void main() {
+  int ans = knapSackProblem(4, 10, [1, 4, 5, 6], [2, 3, 6, 7]);
```
Comment (Parowicz) [inline dynamic_programming/01knapsack_recursive.dart:24]:
```suggestion
  int ans = knapSackProblem(10, [1, 4, 5, 6], [2, 3, 6, 7]);
```

File: dynamic_programming/01knapsack_recursive.dart [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+  print(ans);
+
+  test('TC: 1', () {
+    expect(knapSackProblem(4, 10, [1, 4, 5, 6], [2, 3, 6, 7]), 10);
```
Comment (Parowicz) [inline dynamic_programming/01knapsack_recursive.dart:28]:
```suggestion
    expect(knapSackProblem(10, [1, 4, 5, 6], [2, 3, 6, 7]), 10);
```

File: dynamic_programming/01knapsack_recursive.dart [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+  });
+
+  ans = knapSackProblem(5, 100, [2, 70, 30, 69, 100], [1, 70, 30, 69, 100]);
```
Comment (Parowicz) [inline dynamic_programming/01knapsack_recursive.dart:31]:
```suggestion
  ans = knapSackProblem(100, [2, 70, 30, 69, 100], [1, 70, 30, 69, 100]);
```

File: dynamic_programming/01knapsack_recursive.dart [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+  print(ans);
+
+  test('TC: 2', () {
+    expect(knapSackProblem(5, 100, [2, 70, 30, 69, 100], [1, 70, 30, 69, 100]),
+        101);
```
Comment (Parowicz) [inline dynamic_programming/01knapsack_recursive.dart:36]:
```suggestion
    expect(
        knapSackProblem(100, [2, 70, 30, 69, 100], [1, 70, 30, 69, 100]), 101);
```

**Comment by akashgkrishnan** [comment]
@Parowicz I think the variable names are confusing to you as I'm not actually reducing the size of the values or the weight list.
I'm actually reducing the index from the last
so the values list or the weight list won't change size but the element in consideration from the last is what is changing. so if we rely on values.length the last element will always be the same element that is being considered
ill update the functions with more readable and intuitive variable names and update the documentation for the function and attach the required Links of explanation.

**Comment by Parowicz** [comment]
Yea, but you can for example take length if not provided (as i did in my proposal). 
Or you can create inner function and hide accumulator.

```dart
int knapSackProblem(int capacity, List<int> values, List<int> weights) {
  int aux(
      int numberOfItems, int capacity, List<int> values, List<int> weights) {
    if (numberOfItems == 0 || capacity == 0) {
      return 0;
    }
    int currentValue = values[numberOfItems - 1];
    int currentWeight = weights[numberOfItems - 1];

    if (weights[numberOfItems - 1] <= capacity) {
      return max(
          currentValue +
              aux(numberOfItems - 1, capacity - currentWeight, values, weights),
          aux(numberOfItems - 1, capacity, values, weights));
    } else {
      return aux(numberOfItems - 1, capacity, values, weights);
    }
  }

  return aux(values.length, capacity, values, weights);
}
```

Exposing `numberOfItems` seems like leaky abstraction.

**Comment by mohammadreza490** [comment]
I commited the suggestions because the suggestion was to use the optional parameters if we check for the value not being null. In this case I think the suggestions make the code better. Thank you for your contribution

**Pull request merged**